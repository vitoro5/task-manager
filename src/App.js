import React, { useState } from "react";
import { useSelector, useDispatch } from "react-redux";

import { Creators as Actions } from "./store/ducks/tasks";

import Panel from "./components/Panel";
import Tasks from "./components/Tasks";
import Button from "./components/Button";

import {
  Container,
  OuterContainer,
  PanelHeader,
  InputControl,
  Input,
  Level,
  LevelLeft,
  LevelRight
} from "./styles";

import "./global.css";

import { ReactComponent as Checklist } from "./icons/checklist.svg";
import { ReactComponent as Delete } from "./icons/delete.svg";

const STATUS = {
  FOCUSED: "focused",
  NORMAL: "normal"
};

function App() {
  const tasks = useSelector(({ tasks }) => tasks.data);
  const filter = useSelector(({ tasks }) => tasks.filter);

  const dispatch = useDispatch();

  const [task, setTask] = useState({ value: "", completed: false });
  const [className, setClassName] = useState(STATUS.NORMAL);

  const handleSubmitTask = e => {
    e.preventDefault();

    dispatch(Actions.addTask(task));
    setTask({ ...task, value: "" });
  };

  const renderTasks = () => {
    const filteredTasks = tasks.filter(task => task.completed === filter);

    if (filter === "all" && tasks.length !== 0) {
      return <Panel>{<Tasks filter={filter} />}</Panel>;
    } else if (filteredTasks.length !== 0 && tasks.length !== 0) {
      return <Panel>{<Tasks filter={filter} />}</Panel>;
    }
  };

  const setFilter = filter => {
    dispatch(Actions.setFilter(filter));
  };

  const handleFocus = () => {
    setClassName(STATUS.FOCUSED);
  };

  const handleBlur = () => {
    setClassName(STATUS.NORMAL);
  };

  const handleClickFullWidthButton = () => {
    dispatch(Actions.clearTasks());
  };

  return (
    <Container className="App">
      <OuterContainer>
        <PanelHeader>Task Manager</PanelHeader>
        <Panel>
          <InputControl onSubmit={handleSubmitTask}>
            <Input
              type="text"
              name="test"
              placeholder="Whats needs to be done?"
              onFocus={handleFocus}
              onBlur={handleBlur}
              value={task.value}
              onChange={e => setTask({ ...task, value: e.target.value })}
            />
            <span>
              <Checklist width="22" height="22" className={className} />
            </span>
          </InputControl>
        </Panel>

        {renderTasks()}

        <Panel>
          <Level>
            <LevelLeft>
              <strong>
                {tasks.filter(task => task.completed === false).length}
              </strong>{" "}
              Items left
            </LevelLeft>
            <LevelRight>
              <Button
                onClick={() => setFilter("all")}
                active={filter === "all" && true}
              >
                All
              </Button>
              <Button
                onClick={() => setFilter(false)}
                active={filter === false && true}
              >
                Active
              </Button>
              <Button
                onClick={() => setFilter(true)}
                active={filter === true && true}
              >
                Completed
              </Button>
            </LevelRight>
          </Level>
        </Panel>
        <Panel>
          <Button
            onClick={handleClickFullWidthButton}
            color="#ff3660"
            fullwidth={true}
          >
            <Delete width="20" height="20" />
            Clear
          </Button>
        </Panel>
      </OuterContainer>
    </Container>
  );
}

export default App;
