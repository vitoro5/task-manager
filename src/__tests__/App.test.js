import React from "react";
import { mount } from "enzyme";
import { Provider } from "react-redux";
import { createStore } from "redux";

import App from "../App";
import reducer from "../store/ducks";

describe("<App /> unit test", () => {
  const initialState = {
    tasks: {
      data: [],
      filter: "all"
    }
  };

  const mockStore = createStore(reducer, initialState);

  const getWrapper = () =>
    mount(
      <Provider store={mockStore}>
        <App />
      </Provider>
    );

  it("should renders without crashing", () => {
    getWrapper();
  });

  it("should renders tasks on submit input", () => {
    const wrapper = getWrapper();

    expect(wrapper.exists("ul")).toBeFalsy();
    wrapper.find("input").simulate("change", { target: { value: "testing" } });
    expect(wrapper.find("input").props().value).toEqual("testing");
    wrapper.find("form").simulate("submit");
    expect(wrapper.exists("ul")).toBeTruthy();
  });

  it("should highlight the input icon on focus", () => {
    const wrapper = getWrapper();

    let input = wrapper.find("input");
    expect(input).toMatchSnapshot();
  });
});
