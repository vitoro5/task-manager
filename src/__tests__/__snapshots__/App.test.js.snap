// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<App /> unit test should highlight the input icon on focus 1`] = `
<input
  className="sc-htoDjs kZQDmi"
  name="test"
  onBlur={[Function]}
  onChange={[Function]}
  onFocus={[Function]}
  placeholder="Whats needs to be done?"
  type="text"
  value=""
/>
`;
